### PR TITLE
MONGOCRYPT-392 add HMAC-SHA-256 to libmongocrypt

### PR DESCRIPTION
**Pull request by kevinAlbs**
# Summary
Add HMAC-SHA-256 crypto primitives in libmongocrypt.

Fix a bug in the Windows implementation of `_native_crypto_hmac_sha_512`. An error in BCryptCreateHash would be followed by [an incorrect call to BCryptDestroyHash](https://github.com/mongodb/libmongocrypt/blob/fb853ed4b3d1fc911f9533adba3ee6490e209000/src/crypto/cng.c#L274). This is resolved by [returning immediately on error](https://github.com/mongodb/libmongocrypt/pull/259/files#diff-16dc3c7e14f64e920340a09b92969cd77b9ae5296b715590b90a5fd93659ee08R286).

# Background & Motivation
HMAC-SHA-256 is required by MONGOCRYPT-392. Token generation requires HMAC-SHA-256.

# Rejected Alternatives
I considered making [kms_sha256_hmac](https://github.com/mongodb/libmongocrypt/blob/fb853ed4b3d1fc911f9533adba3ee6490e209000/kms-message/src/kms_crypto.h#L53-L59) accessible to libmongocrypt. KMS message is used to generate KMS requests. Exposing crypto primitives seems out of scope of KMS message. The HMAC functions are short to implement. So I decided to implement HMAC-SHA-256 in libmongocrypt.

### CODE REVIEW

File: src/mongocrypt-crypto.c [hunk 1]
```diff
@@ -991,3 +989,33 @@ _mongocrypt_unwrap_key (_mongocrypt_crypto_t *crypto,
...
+bool
+_mongocrypt_hmac_sha_256 (_mongocrypt_crypto_t *crypto,
+                          const _mongocrypt_buffer_t *key,
+                          const _mongocrypt_buffer_t *in,
+                          _mongocrypt_buffer_t *out,
+                          mongocrypt_status_t *status)
+{
+   if (key->len != MONGOCRYPT_MAC_KEY_LEN) {
+      CLIENT_ERR ("invalid hmac_sha_256 key length. Got %" PRIu32
+                  " expect: %" PRIu32,
```
Comment (eramongodb) [inline src/mongocrypt-crypto.c:1002]:
```suggestion
      CLIENT_ERR ("invalid hmac_sha_256 key length. Got %" PRIu32
                  ", expected: %" PRIu32,
```

File: src/mongocrypt-crypto.c [hunk 2]
```diff
@@ -991,3 +989,33 @@ _mongocrypt_unwrap_key (_mongocrypt_crypto_t *crypto,
...
+      mongocrypt_binary_t key_bin, out_bin, in_bin;
+      bool ret;
+
+      _mongocrypt_buffer_to_binary (key, &key_bin);
+      _mongocrypt_buffer_to_binary (out, &out_bin);
+      _mongocrypt_buffer_to_binary (in, &in_bin);
+
+      ret = crypto->hmac_sha_256 (
+         crypto->ctx, &key_bin, &in_bin, &out_bin, status);
+      return ret;
```
Comment (eramongodb) [inline src/mongocrypt-crypto.c:1018]:
```suggestion
      _mongocrypt_buffer_to_binary (key, &key_bin);
      _mongocrypt_buffer_to_binary (out, &out_bin);
      _mongocrypt_buffer_to_binary (in, &in_bin);

      return crypto->hmac_sha_256 (
         crypto->ctx, &key_bin, &in_bin, &out_bin, status);
```
Omit redundant local variable `ret`.

File: src/mongocrypt-crypto.c [hunk 3]
```diff
@@ -991,3 +989,33 @@ _mongocrypt_unwrap_key (_mongocrypt_crypto_t *crypto,
...
+   }
+   return _native_crypto_hmac_sha_256 (key, in, out, status);
+}
```
Comment (eramongodb) [inline src/mongocrypt-crypto.c:1021]:
```suggestion
}

```
Missing EOF newline.

**Review by eramongodb — APPROVED** [review]
Minor feedback; otherwise, LGTM.

**Comment by kevinAlbs** [comment]
@eramongodb you can dismiss the review request. That was accidental.

**Review by eramongodb — APPROVED** [review]

**Review by sgolemon — APPROVED** [review]
On first read of your commit message I thought you were literally implementing HMAC from scratch, but this is just wrapping calls into native crypto (and they all look valid from a quick glance at the source docs).

LGTM

File: test/test-mongocrypt-crypto.c [hunk 4]
```diff
@@ -333,11 +333,143 @@ _test_native_crypto_aes_256_ctr (_mongocrypt_tester_t *tester)
...
+   const char *testname;
+   const char *key;
+   const char *input;
+   const char *expect;
+} hmac_sha_256_test_t;
+
+void
+_test_native_crypto_hmac_sha_256 (_mongocrypt_tester_t *tester)
+{
+   /* Test data generated with OpenSSL CLI:
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply test/test-mongocrypt-crypto.c:346]: Oh, thank you for the suggestion. I was not aware of those test vectors. Added as part of 35341256cc4b4b8a334bfd34ffbabb96a78b0ff9 and will merge when tests have passed.

**Pull request merged**